File: haskell-server/src/EagerWords/Common/Message/CommitPlayResponse.hs

--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}

module EagerWords.Common.Message.CommitPlayResponse (
    CommitPlayResponse(..)
  , tupleToCommitPlayResponse
  ) where

import GHC.Generics (Generic)
import Data.Aeson (FromJSON, ToJSON)
import Control.DeepSeq (NFData)

import Bolour.Plane.Domain.Point
import EagerWords.Common.Domain.Piece
import EagerWords.Common.Domain.GameMiniState

data CommitPlayResponse = CommitPlayResponse {
    gameMiniState :: GameMiniState
  , replacementPieces :: [Piece]
  , deadPoints :: [Point]
}
  deriving (Eq, Show, Generic, NFData)

instance FromJSON CommitPlayResponse
instance ToJSON CommitPlayResponse

tupleToCommitPlayResponse :: (GameMiniState, [Piece], [Point]) -> CommitPlayResponse
tupleToCommitPlayResponse (miniState, replacementPieces, deadPoints) =
  CommitPlayResponse miniState replacementPieces deadPoints

File: scala-server/app/com/bolour/eagerwords/common/message/CommitPlayResponse.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */

package com.bolour.eagerwords.common.message

import com.bolour.eagerwords.common.domain.{GameMiniState, Piece}
import com.bolour.plane.domain.Point

/**
  * API response to a request to commit a play by the user.
  *
  * @param gameMiniState Mini-state of the game after the play.
  * @param replacementPieces Pieces to replace the user's played pieces.
  * @param deadPoints Empty points on the board that have been determined to be
  *                   no longer playable after the user's play was laid out on the board.
  */
case class CommitPlayResponse(
  gameMiniState: GameMiniState,
  replacementPieces: List[Piece],
  deadPoints: List[Point]
)
